File: cpp-core/src/nmea.cpp
#include "nmea.hpp"

#include <algorithm>
#include <cctype>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <stdexcept>

namespace krot {
namespace {

std::string trim(const std::string& value) {
  auto begin = value.begin();
  while (begin != value.end() && std::isspace(static_cast<unsigned char>(*begin))) {
    ++begin;
  }
  auto end = value.end();
  while (end != begin && std::isspace(static_cast<unsigned char>(*(end - 1)))) {
    --end;
  }
  return std::string(begin, end);
}

std::vector<std::string> split(const std::string& value, char delimiter) {
  std::vector<std::string> fields;
  std::stringstream stream(value);
  std::string item;
  while (std::getline(stream, item, delimiter)) {
    fields.push_back(item);
  }
  if (!value.empty() && value.back() == delimiter) {
    fields.emplace_back();
  }
  return fields;
}

double parse_nmea_time(const std::string& value) {
  if (value.size() < 6) {
    return 0.0;
  }

  try {
    const double hours = std::stod(value.substr(0, 2));
    const double minutes = std::stod(value.substr(2, 2));
    const double seconds = std::stod(value.substr(4));
    return hours * 3600.0 + minutes * 60.0 + seconds;
  } catch (...) {
    return 0.0;
  }
}

std::string uppercase(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
    return static_cast<char>(std::toupper(c));
  });
  return value;
}

}  // namespace

std::uint8_t nmea_checksum(const std::string& payload) {
  std::uint8_t checksum = 0;
  for (const unsigned char item : payload) {
    checksum ^= item;
  }
  return checksum;
}

std::string checksum_hex(std::uint8_t value) {
  std::ostringstream stream;
  stream << std::uppercase << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(value);
  return stream.str();
}

std::optional<RadioAltimeterSample> parse_gga_sentence(const std::string& sentence) {
  const std::string trimmed = trim(sentence);
  if (trimmed.empty()) {
    return std::nullopt;
  }

  std::string body = trimmed;
  if (!body.empty() && body.front() == '$') {
    body.erase(body.begin());
  }

  const std::size_t checksum_separator = body.find('*');
  const std::string payload = checksum_separator == std::string::npos
      ? body
      : body.substr(0, checksum_separator);
  const std::optional<std::string> provided_checksum = checksum_separator == std::string::npos
      ? std::nullopt
      : std::optional<std::string>(uppercase(trim(body.substr(checksum_separator + 1))));

  const std::vector<std::string> fields = split(payload, ',');
  if (fields.empty() || (fields[0] != "GPGGA" && fields[0] != "GNGGA")) {
    return std::nullopt;
  }
  if (fields.size() <= 9) {
    return std::nullopt;
  }

  double radio_altitude_m = 0.0;
  try {
    radio_altitude_m = std::stod(fields[9]);
  } catch (...) {
    return std::nullopt;
  }

  const std::string expected_checksum = checksum_hex(nmea_checksum(payload));
  ChecksumStatus status = ChecksumStatus::Missing;
  if (provided_checksum.has_value()) {
    status = *provided_checksum == expected_checksum ? ChecksumStatus::Ok : ChecksumStatus::Invalid;
  }

  return RadioAltimeterSample{
      parse_nmea_time(fields.size() > 1 ? fields[1] : "0"),
      radio_altitude_m,
      trimmed,
      status,
  };
}

NmeaParseReport parse_nmea_stream(const std::string& text) {
  NmeaParseReport report;
  std::stringstream stream(text);
  std::string line;

  while (std::getline(stream, line)) {
    if (trim(line).empty()) {
      continue;
    }

    const std::optional<RadioAltimeterSample> sample = parse_gga_sentence(line);
    if (!sample.has_value()) {
      report.skipped_lines += 1;
      continue;
    }

    switch (sample->checksum_status) {
      case ChecksumStatus::Ok:
        report.checksum_ok += 1;
        break;
      case ChecksumStatus::Missing:
        report.checksum_missing += 1;
        break;
      case ChecksumStatus::Invalid:
        report.checksum_invalid += 1;
        break;
    }
    report.samples.push_back(*sample);
  }

  return report;
}

NmeaParseReport read_nmea_file(const std::string& path) {
  std::ifstream file(path);
  if (!file) {
    throw std::runtime_error("не удалось открыть журнал: строка радиовысотомера недоступна");
  }

  std::stringstream buffer;
  buffer << file.rdbuf();
  return parse_nmea_stream(buffer.str());
}

}  // namespace krot

File: rust-core/src/nmea.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChecksumStatus {
    Ok,
    Missing,
    Invalid,
}

#[derive(Debug, Clone)]
pub struct RadioAltimeterSample {
    pub time_s: f64,
    pub radio_altitude_m: f64,
    pub sentence: String,
    pub checksum_status: ChecksumStatus,
}

#[derive(Debug, Clone, Default)]
pub struct NmeaParseReport {
    pub samples: Vec<RadioAltimeterSample>,
    pub skipped_lines: usize,
    pub checksum_ok: usize,
    pub checksum_missing: usize,
    pub checksum_invalid: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NmeaError {
    Empty,
    UnsupportedSentence(String),
    MissingRadioAltitude,
}

pub fn nmea_checksum(payload: &str) -> u8 {
    payload.bytes().fold(0_u8, |acc, byte| acc ^ byte)
}

fn checksum_hex(value: u8) -> String {
    format!("{value:02X}")
}

fn parse_nmea_time(value: &str) -> f64 {
    if value.len() < 6 {
        return 0.0;
    }

    let hours = value[0..2].parse::<f64>().unwrap_or(0.0);
    let minutes = value[2..4].parse::<f64>().unwrap_or(0.0);
    let seconds = value[4..].parse::<f64>().unwrap_or(0.0);
    hours * 3600.0 + minutes * 60.0 + seconds
}

pub fn parse_gga_sentence(sentence: &str) -> Result<RadioAltimeterSample, NmeaError> {
    let trimmed = sentence.trim();
    if trimmed.is_empty() {
        return Err(NmeaError::Empty);
    }

    let body = trimmed.strip_prefix('$').unwrap_or(trimmed);
    let (payload, provided_checksum) = match body.split_once('*') {
        Some((payload, checksum)) => (payload, Some(checksum.trim().to_uppercase())),
        None => (body, None),
    };
    let fields: Vec<&str> = payload.split(',').collect();
    let sentence_type = fields.first().copied().unwrap_or_default();

    if sentence_type != "GPGGA" && sentence_type != "GNGGA" {
        return Err(NmeaError::UnsupportedSentence(sentence_type.to_string()));
    }

    let radio_altitude_m = fields
        .get(9)
        .and_then(|value| value.parse::<f64>().ok())
        .ok_or(NmeaError::MissingRadioAltitude)?;

    let expected = checksum_hex(nmea_checksum(payload));
    let checksum_status = match provided_checksum {
        Some(value) if value == expected => ChecksumStatus::Ok,
        Some(_) => ChecksumStatus::Invalid,
        None => ChecksumStatus::Missing,
    };

    Ok(RadioAltimeterSample {
        time_s: parse_nmea_time(fields.get(1).copied().unwrap_or("0")),
        radio_altitude_m,
        sentence: trimmed.to_string(),
        checksum_status,
    })
}

pub fn parse_nmea_stream(text: &str) -> NmeaParseReport {
    let mut report = NmeaParseReport::default();

    for line in text.lines().map(str::trim).filter(|line| !line.is_empty()) {
        match parse_gga_sentence(line) {
            Ok(sample) => {
                match sample.checksum_status {
                    ChecksumStatus::Ok => report.checksum_ok += 1,
                    ChecksumStatus::Missing => report.checksum_missing += 1,
                    ChecksumStatus::Invalid => report.checksum_invalid += 1,
                }
                report.samples.push(sample);
            }
            Err(_) => {
                report.skipped_lines += 1;
            }
        }
    }

    report
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_case_shape_and_marks_checksum() {
        let sample = parse_gga_sentence("$GPGGA,123519.111,,,,,,,,545.4,M,46.9,M,,*47").unwrap();
        assert_eq!(sample.radio_altitude_m, 545.4);
        assert_eq!(sample.checksum_status, ChecksumStatus::Invalid);
        assert_eq!(
            checksum_hex(nmea_checksum("GPGGA,123519.111,,,,,,,,545.4,M,46.9,M,,")),
            "7F"
        );
    }
}
